src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Net.Mime;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    public static class HttpContentJsonExtensions
+    {
+        public static Task<object?> ReadFromJsonAsync(this HttpContent content, Type type, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+        {
+            return ReadFromJsonAsyncCore(content, type, options, cancellationToken);
+        }
+
+        public static Task<T> ReadFromJsonAsync<T>(this HttpContent content, JsonSerializerOptions? options = null, CancellationToken cancellationToken = default)
+        {
+            return ReadFromJsonAsyncCore<T>(content, options, cancellationToken);
+        }
+
+        private static async Task<object?> ReadFromJsonAsyncCore(HttpContent content, Type type, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            byte[] contentBytes = await GetUtf8JsonBytesFromContentAsync(content, cancellationToken).ConfigureAwait(false);
+            return JsonSerializer.Deserialize(contentBytes, type, options);
```

Comment:
I wanted to use `DeserializeAsync` and `ReadAsStreamAsync`, but if we need to transcode from anything to `UTF-8`, I think we need the whole content, so it makes more sense to use `ReadAsByteArrayAsync`.

In addition, `cancellationToken` is useless since `ReadAsByteArrayAsync` does not take such argument.

Reply:
> if we need to transcode from anything to UTF-8, I think we need the whole content

You can avoid this. You should use this API once it gets implemented: https://github.com/dotnet/runtime/issues/30260

Reply:
> In addition, `cancellationToken` is useless since `ReadAsByteArrayAsync` does not take such argument.

We do have a cancellable overload, just not in .NET Standard. We should dual-target the library.

Reply:
This logic needs to use the Stream API, it can be `byte[]` for non utf8 which isn't the common case.